src/pal/src/arch/arm/context2.S

```diff
@@ -0,0 +1,203 @@
+#define CONTEXT_D26          CONTEXT_D25+8
+#define CONTEXT_D27          CONTEXT_D26+8
+#define CONTEXT_D28          CONTEXT_D27+8
+#define CONTEXT_D29          CONTEXT_D28+8
+#define CONTEXT_D30          CONTEXT_D29+8
+#define CONTEXT_D31          CONTEXT_D30+8
+
+// Incoming:
+//  r0: Context*
+//
+LEAF_ENTRY CONTEXT_CaptureContext, _TEXT
+    // Ensure we save these registers
+    push {r4-r11}
+    // Save processor flags before calling any of the following 'test' instructions
+    // because they will modify state of some flags
+    push {r1}
+    mrs r1, apsr // Get APSR - equivalent to eflags
+    push {r1} // Save APSR
+    END_PROLOGUE
+    
+    push {r2}
+    ldr r2, [R0, #(CONTEXT_ContextFlags)]
+    tst r2, #(CONTEXT_INTEGER)
+    pop {r2}
+    
+    // Add 4 to stack so we point at R1, pop, then sub 8 to point at APSR
+    add sp, sp, #4
+    pop {r1}
+    sub sp, sp, #8
+    
+    strne r0, [r0, #(CONTEXT_R0)]
+    addne r0, CONTEXT_R1
+    stmiane r0, {r1-r12}
+    subne r0, CONTEXT_R1
+    
+    ldr r2, [r0, #(CONTEXT_ContextFlags)]
+    tst r2, #(CONTEXT_CONTROL)
+    
+    addne sp, sp, #(10*4) // This needs to put the stack in the same state as it started
+    strne sp, [r0, #(CONTEXT_Sp)]
+    subne sp, sp, #(10*4)
+    
+    strne lr, [r0, #(CONTEXT_Lr)]
+    strne lr, [r0, #(CONTEXT_Pc)]
+    
+    // Get the APSR pushed onto the stack at the start
+    pop {r1}
+    strne r1, [r0, #(CONTEXT_Cpsr)]
+    
+    ldr r2, [r0, #(CONTEXT_ContextFlags)]
+    tst r2, #(CONTEXT_FLOATING_POINT)
+    
+    vmrsne r3, fpscr
+    strne r3, [r0, #(CONTEXT_Fpscr)]
+    
+    addne r0, CONTEXT_D0
+    vstmiane r0!, {d0-d15}
+    vstmiane r0!, {d16-d31}
+    subne r0, CONTEXT_D31
+    
+    // Make sure sp is restored
+    add sp, sp, #4
+
+    // Restore callee saved registers
+    pop {r4-r11}
+    bx lr
+LEAF_END CONTEXT_CaptureContext, _TEXT
+
+// Incoming:
+//  R0: Context*
+//
+LEAF_ENTRY RtlCaptureContext, _TEXT
+    push {r1}
+    mov r1, #0
+    orr r1, r1, #CONTEXT_ARM
+    orr r1, r1, #CONTEXT_INTEGER
+    orr r1, r1, #CONTEXT_CONTROL
+    orr r1, r1, #CONTEXT_FLOATING_POINT
+    str r1, [r0, #(CONTEXT_ContextFlags)]
+    pop {r1}
+    b C_FUNC(CONTEXT_CaptureContext)
+LEAF_END RtlCaptureContext, _TEXT
+
+// Incoming:
+//  r0: Context*
+//  r1: Exception*
+//
+LEAF_ENTRY RtlRestoreContext, _TEXT
+    END_PROLOGUE
+    
+    ldr r2, [r0, #(CONTEXT_ContextFlags)]
+    tst r2, #(CONTEXT_FLOATING_POINT)
+    
+    addne r0, CONTEXT_D0
+    vldmiane r0!, {d0-d15}
+    vldmiane r0, {d16-d31}
+    subne r0, CONTEXT_D16
+    
+    ldrne r3, [r0, #(CONTEXT_Fpscr)]
+    vmrsne r3, FPSCR
+    
+    ldr r2, [r0, #(CONTEXT_ContextFlags)]
+    tst r2, #(CONTEXT_CONTROL)
+    
+    beq LOCAL_LABEL(No_Restore_CONTEXT_CONTROL)
+    
+    ldr r2, [r0, #(CONTEXT_ContextFlags)]
+    tst r2, #(CONTEXT_INTEGER)
+    
+    beq LOCAL_LABEL(No_Restore_CONTEXT_INTEGER)
+    
+    ldr R2, [r0, #(CONTEXT_Cpsr)]
+    msr APSR, r2    
```

Comment:
The jump must occur before restoring APSR otherwise the flag that the conditional depends upon could change. So APSR is restored immediately after either result of the branch.